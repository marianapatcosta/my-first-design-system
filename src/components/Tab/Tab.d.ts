import React, { CSSProperties, MouseEventHandler } from 'react';
export interface TabProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    tabClassName?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
    /**
     * Dropdown's label
     */
    label: string;
    /**
     * Is Tab disabled?
     */
    disabled?: boolean;
    /**
     * Is Tab active, i.e, is the one is content is being rendered?
     */
    isActive?: boolean;
    /**
     * Click handler
     */
    onClick: MouseEventHandler<HTMLLIElement>;
}
export declare const Tab: React.FC<TabProps>;
