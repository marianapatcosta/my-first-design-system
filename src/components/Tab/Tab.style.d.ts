export declare const StyledTab: import("styled-components").StyledComponent<"li", any, {
    isActive?: boolean | undefined;
    disabled?: boolean | undefined;
}, never>;
