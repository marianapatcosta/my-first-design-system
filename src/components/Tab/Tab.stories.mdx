import { Meta } from "@storybook/addon-docs/blocks";
import { Tab } from "./Tab";

<Meta title="Design System|Tab" component={Tab} />

# Tab
Tab component to show a certain content that will be rendered when the tab is active

## Usage
Tab should be used as child of 'Tabs' component and its content is only rendered when the tab is active

- Default Tab:
  <Story name='Default'>
    <Tab label='Tab 1'/>
  </Story>

- Active Tab:
  <Story name='Active'>
    <Tab label='Tab 1' isActive={true}/>
  </Story>

## Other important notes
not applicable.