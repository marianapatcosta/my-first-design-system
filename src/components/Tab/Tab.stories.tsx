import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Tab, TabProps } from './Tab';

export default {
  title: 'Tab',
  component: Tab,
  // args behavior in each story
  args: { label: 'Tab 1' },
} as Meta;

const Template: Story<TabProps> = (args) => <Tab {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const Active = Template.bind({});
Active.args = { isActive: true };
