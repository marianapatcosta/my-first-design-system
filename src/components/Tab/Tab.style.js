var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';
export var StyledTab = styled.li(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", ";\n  min-width: 3.125rem;\n  max-width: 4rem;\n  color: ", ";\n  padding: 0.3rem 0.625rem;\n  text-align: center;\n  position: relative;\n  user-select: none;\n  cursor: pointer;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  ", "\n\n  ", "\n"], ["\n  ", ";\n  min-width: 3.125rem;\n  max-width: 4rem;\n  color: ", ";\n  padding: 0.3rem 0.625rem;\n  text-align: center;\n  position: relative;\n  user-select: none;\n  cursor: pointer;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  ",
    "\n\n  ",
    "\n"])), baseStyles, colors.highlight, function (_a) {
    var disabled = _a.disabled;
    return disabled &&
        "\n    " + disabledStyles + "\n    ";
}, function (_a) {
    var isActive = _a.isActive;
    return isActive &&
        "font-weight: 700; \n        background-color: " + colors.primary + ";\n        box-shadow: 0 0 0.125rem " + colors.shadow + ";\n        border-top-left-radius: 0.625rem;\n        border-top-right-radius: 0.625rem;\n\n        :before {\n          content: \"\";\n          position: absolute;\n          width: 100%;\n          height: 0.3rem;\n          bottom: -0.3rem;\n          left: 0;\n          background-color: " + colors.primary + ";\n        }\n      ";
});
var templateObject_1;
