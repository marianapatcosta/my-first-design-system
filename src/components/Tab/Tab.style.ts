import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';

export const StyledTab = styled.li<{ isActive?: boolean; disabled?: boolean }>`
  ${baseStyles};
  min-width: 3.125rem;
  max-width: 4rem;
  color: ${colors.highlight};
  padding: 0.3rem 0.625rem;
  text-align: center;
  position: relative;
  user-select: none;
  cursor: pointer;

  :hover {
    opacity: 0.75;
  }

  ${({ disabled }) =>
    disabled &&
    `
    ${disabledStyles}
    `}

  ${({ isActive }) =>
    isActive &&
    `font-weight: 700; 
        background-color: ${colors.primary};
        box-shadow: 0 0 0.125rem ${colors.shadow};
        border-top-left-radius: 0.625rem;
        border-top-right-radius: 0.625rem;

        :before {
          content: "";
          position: absolute;
          width: 100%;
          height: 0.3rem;
          bottom: -0.3rem;
          left: 0;
          background-color: ${colors.primary};
        }
      `}
`;
