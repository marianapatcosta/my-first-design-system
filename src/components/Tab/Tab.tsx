import React, { CSSProperties, MouseEventHandler } from 'react';
import { isEventValid } from '../../utils/utils';
import { StyledTab } from './Tab.style';

export interface TabProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  tabClassName?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
  /**
   * Dropdown's label
   */
  label: string;
  /**
   * Is Tab disabled?
   */
  disabled?: boolean;
  /**
   * Is Tab active, i.e, is the one is content is being rendered?
   */
  isActive?: boolean;
  /**
   * Click handler
   */
  onClick: MouseEventHandler<HTMLLIElement>;
}

export const Tab: React.FC<TabProps> = ({
  disabled,
  isActive,
  label,
  onClick,
  tabClassName,
  ...props
}) => {
  const handleTabSelection = (event: any) =>
    isEventValid(event) && !disabled && onClick(event);

  return (
    <StyledTab
      role='tab'
      aria-selected={isActive}
      tabIndex={disabled ? -1 : 0}
      id={label}
      className={tabClassName}
      isActive={isActive}
      disabled={disabled}
      onClick={handleTabSelection}
      onKeyDown={handleTabSelection}
      {...props}
    >
      {label}
    </StyledTab>
  );
};

Tab.defaultProps = {
  id: '',
  tabClassName: '',
  style: {},
  label: '',
  disabled: false,
  isActive: false,
  onClick: () => null,
};
