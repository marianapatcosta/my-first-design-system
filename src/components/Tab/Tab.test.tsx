import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Tab } from './Tab';

describe('Tab', () => {
  const defaultProps = {
    label: 'Tab 1',
    onClick: () => null,
  };

  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Tab {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should handle click events', () => {
    const onClickMock = jest.fn();
    const {
      container: { firstChild },
    } = render(<Tab {...defaultProps} onClick={onClickMock} />);

    firstChild && fireEvent.click(firstChild);

    expect(onClickMock).toBeCalled();
  });

  describe('Disabled Tab', () => {
    test('should display label', () => {
      const { getByText } = render(<Tab {...defaultProps} disabled />);

      getByText('Tab 1');
    });

    test('should return null on click if onClick prop is not passed', () => {
      const result = Tab.defaultProps?.onClick && Tab.defaultProps.onClick();
      expect(result).toBeNull();
    });

    test('should be disabled', () => {
      const {
        container: { firstChild },
      } = render(<Tab {...defaultProps} disabled={true} />);

      expect(firstChild).toHaveAttribute('disabled');
    });

    test('should not handle click events', () => {
      const onClickMock = jest.fn();
      const {
        container: { firstChild },
      } = render(
        <Tab {...defaultProps} disabled={true} onClick={onClickMock} />
      );

      firstChild && fireEvent.click(firstChild);

      expect(onClickMock).not.toBeCalled();
    });
  });
});
