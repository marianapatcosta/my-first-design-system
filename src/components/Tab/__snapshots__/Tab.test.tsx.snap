// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab renders without errors and matches snapshot 1`] = `
<li
  aria-selected="false"
  class="sc-bdfBwQ cQWGAO"
  id=""
  role="tab"
  tabindex="0"
>
  Tab 1
</li>
`;
